src/analyses/ai_history.h

```diff
@@ -0,0 +1,237 @@
+
+#include <util/json.h>
+#include <util/xml.h>
+
+/// A history object is an abstraction / representation of the control
+/// control-flow part of a set of traces.  The simplest example is a single
+/// location which represents "all traces that reach this location".  More
+/// sophisticated history objects represent "all traces that reach this point
+/// after exactly N iterations of this loop", "all traces that follow a
+/// given pattern of branching to reach this point" or "all traces that have
+/// the following call stack and reach this location".
+///
+/// These are used to control the abstract interpreter; edges are computed
+/// per history.  Depending on what storage is used, they may also control
+/// or influence the number of domains that are used, supporting delayed
+/// merging, loop unwinding, context dependent analysis, etc.
+
+/// This is the base interface; derive from this.
+class ai_history_baset
+{
+public:
+  typedef goto_programt::const_targett locationt;
+  typedef ai_history_baset tracet;
+
+  /// History objects are intended to be immutable so they can be shared
+  /// to reduce memory overhead
+  typedef std::shared_ptr<const tracet> trace_ptrt;
+
+  /// Create a new history starting from a given location
+  /// This is used to start the analysis from scratch
+  /// PRECONDITION(l.is_dereferenceable());
+  explicit ai_history_baset(locationt)
+  {
+  }
+
+  ai_history_baset(const ai_history_baset &)
+  {
+  }
+
+  virtual ~ai_history_baset()
+  {
+  }
+
+  enum step_status
+  {
+    NEW,
+    MERGED,
+    BLOCKED
+  };
+
+  /// In a number of places we need to work with sets of histories
+  /// so that these (a) make use of the immutability and sharing and
+  /// (b) ownership can be transfered if necessary, we use shared pointers
+  /// rather than references.
+  /// The custom comparison allows particular histories to control
+  /// the order of exploration.
+  struct compare_historyt
+  {
+    bool operator()(const trace_ptrt &l, const trace_ptrt &r) const
+    {
+      return *l < *r;
+    }
+  };
+  typedef std::set<trace_ptrt, compare_historyt> trace_sett;
+
+  /// Step creates a new history by advancing the current one to location "to"
+  /// It is given the set of all other histories that have reached this point.
+  ///
+  /// PRECONDITION(to.id_dereferenceable());
+  /// PRECONDITION(to in goto_program.get_successors(current_location()) ||
+  ///              current_location()->is_function_call() ||
+  ///              current_location()->is_end_function());
+  ///
+  /// Step may do one of three things :
+  ///  1. Create a new history object and return a pointer to it
+  ///     This will force the analyser to continue regardless of domain changes
+  ///     POSTCONDITION(IMPLIES(result.first == NEW,
+  ///                           result.second.use_count() == 1 ));
+  ///  2. Merge with an existing history
+  ///     This will mean the analyser will continue of the domain is modified
```

Comment:
Could you please rephrase this line?

Reply:
"This means the analyser will only continue if the domain is updated"